### PR TITLE
Implement actual ECE buttons in Block editor

### LINKED ISSUE #9417: [ECE] Replace SVGs with ECE buttons in the Blocks Editor preview
### Describe the bug
In the Cart & Checkout Blocks, the Editor preview currently renders ECE buttons as static SVGs instead of interactive, fully styled buttons. This limits the ability to accurately visualize the button layout within the editor.

The new `buttonAttributes` API (integrated with WooPayments in [this PR](https://github.com/Automattic/woocommerce-payments/pull/8888)) was designed to ensure that express checkout buttons are rendered with their correct layout in the Editor preview, reflecting styling adjustments such as `height` and `borderRadius`. 

To achieve this, we need to replace the SVGs currently in use with real ECE buttons, similar to how they are rendered in the WooPayments settings page for express checkout.

### LINKED ISSUE #9465: Editing the Checkout Block displays express payment method buttons even when they are disabled
### Describe the bug

Express payment method buttons are displayed when editing the Checkout Block even when the payment method is disabled in the WooPayments settings.

> [!NOTE]
> This only seems to affect Apple Pay and Google Pay.

### To Reproduce

1. Go to Payments > Settings
2. Disable all express payment methods.
3. Open the Checkout Block page in the editor
4. See that the Apple Pay and/or Google Pay buttons are still rendered in the editor

### Actual behavior

Express payment method buttons are visible in the editor when they are disabled in settings.

### Screenshots

CleanShot 2024-09-18 at 21 12 26

### Expected behavior

Express payment method buttons should not be visible in the editor when they are disabled in settings.

### Potential solution
We should only call `registerExpressPaymentMethod` if it is enabled from the WooPayments settings page

### Additional context
Originally reported for Stripe for WooCommerce: https://github.com/woocommerce/woocommerce-gateway-stripe/issues/3437

### PR DESCRIPTION
**Pull request by cesarcosta99**
Resolves #9417.
Fixes #9465.

#### Changes proposed in this Pull Request

This PR does three things:

- Replaces the implementation of ECE buttons in the Block editor. Previously they were plain SVGs, now they are "non-functional" ECE instances carrying all styling settings defined by the merchant.
- Ensures the ECE buttons display only when they're enabled in WooPayments.
- Adds support for styling controls in the ECE buttons that will become available in WooCommerce soon.

#### Testing instructions

**Setup**

1. Turn on the public-facing HTTPS URL in your site.
2. Build the assets: `npm run build:client`.
3. You should run these tests on Safari to be sure Apple Pay and Google Pay buttons are all displayed, but for additional coverage, would be really nice to run the same tests on Google Chrome as well.

**Regression test: Ensure ECE buttons still work as expected**

1. As a shopper, complete checkout in a product page.
2. Complete checkout on cart and checkout pages for both Blocks and shortcode.

**Bonus regression test: Ensure payment request buttons still work as expected**

1. Turn off [the ECE feature flag](https://github.com/Automattic/woocommerce-payments/blob/49328e8a0dcb00366f03cf110e3947402e62230a/includes/class-wc-payments-features.php#L353).
2. As a shopper, complete checkout in a product page.
3. Complete checkout on cart and checkout pages for both Blocks and shortcode.
4. Turn the ECE feature flag back on.

**Regression test: Ensure ECE buttons still look as expected**

1. As a merchant, navigate to **Payments > Settings**.
2. Scroll down to **Express checkouts** and click on **Customize** for Apple Pay/Google Pay.
5. Play with the styling settings and confirm, as a shopper, in each product/cart/checkout (shortcode and Blocks) pages that the settings are being correctly applied.
6. As a merchant, navigate to **Payments > Settings**.
7. Disable Apple Pay/Google Pay.
8. As a shopper, confirm the buttons are not displayed in each product/cart/checkout (shortcode and Blocks) pages.

**Test: ECE buttons are rendered in the Block editor**

1. As a merchant, navigate to **Pages > All Pages**.
3. Edit the Cart (Blocks) page. Repeat the same for the Checkout (Blocks) page.
4. Confirm the ECE buttons are not displayed as you disabled it in the previous test.
5. Navigate to **Payments > Settings** and re-enable Apple Pay/Google Pay.
7. Go back to editing the Cart (Blocks) page and confirm the ECE buttons are rendered.
9. While in **Payments > Settings**, play with the styling settings and confirm the settings are being correctly applied in the Block preview.
10. Click on the buttons and ensure nothing happens.

**Test: Ensure ECE buttons support styling controls in the Block editor**

> [!TIP]
> Check this [WooCommerce pull request](https://github.com/woocommerce/woocommerce/pull/51296) to have an idea what is this about.

1. As a merchant, navigate to **Pages > All Pages**.
2. Edit the Cart (Blocks) page. Repeat the same for the Checkout (Blocks) page.
3. Click on the ECE button and confirm the styling controls are not displayed.
4. Get the latest WooCommerce  `trunk` branch in your site.
5. Refresh the page and confirm that clicking on the ECE button now gets the styling controls displayed.

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

**Post merge**

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

### CODE REVIEW

**Comment by botwoo** [comment]
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `9650` or branch name `update/9417-ece-in-block-editor` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&woocommerce-payments-dev-tools&branches.woocommerce-payments=update/9417-ece-in-block-editor) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: ce533ce30a047fdce04a867b9af2da312a7379ff
- Build time: 2024-10-31 17:36:52 UTC

_Note: the build is updated when a new commit is pushed to this PR._

File: client/checkout/blocks/index.js
Comment (cesarcosta99) [inline client/checkout/blocks/index.js]: `isExpressCheckoutElementEnabled` just retrieves whether the ECE feature flag is turned on. So I created this `isPaymentRequestEnabled` variable to retrieve whether ApplePay and Google Pay are actually enabled.

In the current implementation, there are inner mechanisms that guarantee the ECE buttons won't be displayed when the setting is disabled even though they're being registered here. With this change, the payment methods won't be registered and the Blocks preview will be able to show the same as in the cart/checkout pages. @rafaelzaleski, can you double-check this change to see if there are any risks I haven't noticed?

File: client/globals.d.ts
Comment (cesarcosta99) [inline client/globals.d.ts]: I set `any` as return here because the `getSetting` function can return pretty much anything. `unkwown` is not really possible as this is now tied to the `WCPayExpressCheckoutParams` interface. Not sure if there would be a better TS approach in this scenario, but sticking with `any` for now.

File: includes/express-checkout/class-wc-payments-express-checkout-button-handler.php
Comment (cesarcosta99) [inline includes/express-checkout/class-wc-payments-express-checkout-button-handler.php:471]: We only need to pass the button settings to the Block editor so it gets all the styling defined in the WooPayments settings.

File: client/globals.d.ts
Comment (reykjalin) [reply client/globals.d.ts]: > [!NOTE]
> **TL;DR:** There's a better way to do this, but I'm not sure how it will interact with the rest of the codebase. Worth trying, but I wouldn't spend too much time on it if it's causing problems.

There is, in fact, a better way to do this, but I don't know if it's appropriate to use here. You can look at [this TS playground](https://www.typescriptlang.org/play/?#code/JYWwDg9gTgLgBAbzgcwKYwMrpsAdsuAXzgDMoIQ4ByKVAQwGMYqBuAKDbxlShMdTgBBEBACuueAjZw4DUVFq4GATwBc1UQGcAJlTgAfagzq72MuiPEx1uUSABGPdoQ5cefBgIDiAGwj26HywYHHxNRGk4CzEJdWEYmDNZAAtUBgBrMRgAIVEQiFwAFVQAD2s4TRgoPGQkiDAcAsCAdWAYZIBVXG1UEjxUbQB+dUrq-DqG4CafVvaAOVEfHxGqmoM4WyXnDgB6HbgAKS14HpFcUbpQgnbgcLpw4BI4AHdUGgFQSFg1sgoKilQz1StDgqB8mlQADo2Hs4AAlehMOC3KIANzowB8dHsPgE9yicDAjHSdDQyNwcBu4TAWOUyHI4m0ABp-nAAFbHZEgUn9SnJFEokTaR7KNYAA2eDGCV00YuhPQYWJBQsWAnejGYERkDAKlRQ6GlawAvHAADwAaVBZVQ3XC6VQyggT18-kChrCLMKVu4trgAAoXQEgtgapoANrmgC660ZvX62gAlAA+P32tRwc0svhLAIZYZwQoJ9SBt0hsIR6OGQrbGH7I31uDNAQ6xbaKLgiBwbSdqkvNrJOBiKCD54Uz64kA2mCXKYUp1wMVoTBl5Bill4F78hgD4wQl4CDl65J0VECbvI+A6hRpGA+ZSQuD1k17Xb7OYQHCePmXOAl4MhUM4G5ZQNg-OBHC7ONcAGaFYR1c54GQPwg3dZBNGLZDSwAsIY26KCBkfOBYz6aDtHYWtZF1RCDRXQiLW9G1tDtB15z-VDNE9BjfQDTD-xlCtcJ6EiBmTVMHXUTNSECHxc3SfNCww10+NDASq0fJMIlhGRHn9FBePYwZITDNNowTTT9hkGRaBgeQKSQpT2OMh1I3YLSiFfSydL9KSc2JOAzKkNyrPQWyfJkvz8WrCiZBcaK4FhJsNlQAj2gEMVIrFSllDAPFNAhb4CjgYEBBgTt7VQMA+QEazQpgbLm2gWgmFgiy4BqqAKWI+MonCKLYVi2EnzgABRbpB3kLkaVQScJBnQrSgsKaHyGl8PPfMLZLgdEfFEVBwknOhzgXLrSMylFDsHSYCmheC9WiKxCKXVC-Soe6JCoBN2Fuy9UgyLJcnyIpSngE0npXF7tzSTI8gB0qgbKD6vqoy7GlwQIuiEx6aOw5AXvqVGWn7DH8N0T62G+lHZ0CBYlix5ccbxq60ZmfsaZ8RG4A8gAxaTNu23bwjRhQIGeKqspyyke1SSjryRIkoAsdAeC2wIBZauB30-ARkhFqqQToEFcE7E6CP5vaip4VByeRt6YFmZIed8jI6ee17LHelkkDka8lHTKhjF0FlbfUABmAAGCOiDJinIb+mG8jh4oyntx3wud0HsauCHfuhnIE4KJPmBZKgAAU6BAo3nkRtaPxK494DFTYfDO8JKkxHwCW24A20YK9hXwSWxbqiXNGUSppvAtI6C0NLIXxqmWfmRZm9kC6INKGlgAYNo7wqdBB8b5e5Qowp+QeO4mFEaSQJ1U9aDbDdUsmicpzmilDrbYxcCoeBjzAHLcAsnsHkOAABJIqJ4bQ-1IMAZA8gCJZCKrrUqQE6D2iHvVUe49KC0BILiJE7RLjq1ARSDwMAWSEPgCAWByR4BG3gBBUQlUUFP2HgILB3BKCOGMDPC8VBwhTmALQXeXc2woPahSQ+SxMrFXJOPEwg4nhP2zGndIKsdpW1hDpSh-DNw-ifuOaar9UYvEOjAcIKDuw3WRvPaYbMU68z8hnemWcqC2OZvbNmVBi70IALRN2rrCDAisoj-x8MAc2KCxQm20DIy2PUFxN0ytiCAp5rEIUptMYmwltAOKdmo5xrt3GE3aNk+M3jqB+JidXCiAARCA5t6EvGgOkB8XQdQgBmvAFBEI64CB4OQKAmh1YU3bqgZAgQXbgyoN2Pa-iPy+NKLcZgZM4LIzGRMjuhTpm2wqfw1Y+B5kwF8b3Cq3BSauX2KMpY4zJnbIZv7HO-187wyLogQc9g2Q3nQtQA2AgmmnIaAMPQhBVlXPWTczZUyHnFMXskLxLIQ5kyAA) I set up for some inspiration on the precise implementation. I'll try to give an overview here and would defer to you on which is better; using `any` or narrowing the function signature.

So if we want to narrow the type here we need 3 things:

1. Define which `setting` string is associated with what type of object;
2. Define the type of object returned for each `setting` string; and then
3. Narrow the function signature to reflect that.

Based on what I see in the PR I think that would look something like this:

```ts
interface WcSettings {
  // I made this optional since it probably won't be loaded every time.
  ece_data?: WCPayExpressCheckoutParams;

  // Other things can be added here in the future to expand the type.
}

const wcSettingsModule: {
  getSettings: <K extends keyof WcSettings, T extends WcSettings[K] | undefined>( setting: K, fallback?: T ): WcSettings[K] | T;
};
```

It's best to refer to the TS playground to make sense of the type signature. The short version is that we make sure to only accept known, valid keys to the function, and then use that to infer valid types for the fallback value and possible return values.

This would give us a lot of type safety, but only in the context of what we define ourselves. It's also possible this would conflict with other uses of the function across the codebase, so if doing this causes issues I'd recommend just using `any`. `any` is mostly fine here, even if the narrowed interface would be a good improvement for existing and future code.

It just might require a lot more work to implement. Worth trying though!

File: changelog/update-9417-ece-in-block-editor
Comment (reykjalin) [inline changelog/update-9417-ece-in-block-editor]: > [!NOTE]
> This is a nitpick.

```suggestion
Use preview versions of the actual ECE buttons in the Block editor.
```

File: includes/express-checkout/class-wc-payments-express-checkout-button-handler.php
Comment (reykjalin) [inline includes/express-checkout/class-wc-payments-express-checkout-button-handler.php]: Should we rename this to `register_ece_data_for_block_editor` so it's clear that this shouldn't be loaded anywhere else?

File: client/express-checkout/blocks/index.js
Comment (reykjalin) [inline client/express-checkout/blocks/index.js:73]: I tried commenting this out and it didn't change anything for me. Isn't this for the new Express Checkout settings API? I'm not sure we actually support that yet, but I'll admit I haven't been following the work on that for a long time 🤔 

Maybe I need to install Gutenberg or something to see the difference?

**Review by reykjalin — CHANGES_REQUESTED** [review]
Tests well in the editor across both Arc (Chrome) and Safari, with ECE enabled and disabled 🎉 

And I'm able to make payments too, with both Apple Pay and Google Pay, each with both ECE enabled and disabled! 🚀 

Looking good and testing great! I did have a couple of comments on the code though, so let's resolve those before I approve :)

File: client/globals.d.ts
Comment (cesarcosta99) [reply client/globals.d.ts]: >we make sure to only accept known, valid keys to the function, and then use that to infer valid types for the fallback value and possible return values

I think the benefit in changing to your suggested approach is what you described here. Changed it in 38da8e0.

File: includes/express-checkout/class-wc-payments-express-checkout-button-handler.php
Comment (cesarcosta99) [reply includes/express-checkout/class-wc-payments-express-checkout-button-handler.php]: Renamed it to `register_ece_data_for_block_editor` in cb0772b.

File: client/express-checkout/blocks/index.js
Comment (cesarcosta99) [reply client/express-checkout/blocks/index.js:73]: Just to confirm, did you got the latest WooCommerce `trunk` branch in your site as described in the 4th step of **Test: Ensure ECE buttons support styling controls in the Block editor**? It looks like it worked for the WooPay button in #9473. You can check https://github.com/woocommerce/woocommerce/pull/51296 to have an idea how it should look like.

If you can't test it for some reason, I think that's fine as this doesn't impact the main purposes of this PR, and the lack of support in the current WooCommerce version doesn't make it of an issue either. Honestly, I wasn't able to test myself because the WooCommerce build command was failing for me, but I figured it should work out of the box when the support comes to the stable WC version.

**Comment by cesarcosta99** [comment]
@reykjalin, FYI, I reverted 206821c because a similar approach has been merged to `develop` via #9646 to fix the same issue (duplicated).

File: client/express-checkout/blocks/index.js
Comment (reykjalin) [reply client/express-checkout/blocks/index.js:73]: Hmm ok ok, I'll see if I can test with WC `trunk` and see if that works!

File: client/checkout/blocks/index.js
Comment (rafaelzaleski) [reply client/checkout/blocks/index.js]: I don't see any risks, and thank you for making this change! Just a minor suggestion: we should [remove the PRB implementation](https://github.com/Automattic/woocommerce-payments/issues/8950) now that ECE has been around for a while. It might be better to use a new name for this flag, so we won’t need to update it again during the code clean-up.

File: client/express-checkout/blocks/index.js
Comment (reykjalin) [reply client/express-checkout/blocks/index.js:73]: Built WC `trunk` and can confirm this works as expected 👍

**Review by reykjalin — APPROVED** [review]
Looks great, tests even better 🎉 

🚀

**Pull request merged**
